src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -21,17 +21,589 @@
+
+    types[0] = MPI_UNSIGNED_LONG_LONG;
+    types[1] = MPI_INT;
+    disps[0] = 0;
+    disps[1] = (MPI_Aint) ((uintptr_t) & maxloc.loc - (uintptr_t) & maxloc.sz);
+
+    mpi_errno = MPIR_Type_create_struct_impl(2, blocks, disps, types, &maxloc_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIR_Type_commit_impl(&maxloc_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIR_Op_create_impl(ull_maxloc_op_func, 0, &maxloc_op);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    maxloc.sz = (unsigned long long) mysz;
+    maxloc.loc = myloc;
+
+    mpi_errno = MPIR_Allreduce(&maxloc, &maxloc_result, 1, maxloc_type, maxloc_op, comm, &errflag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    *maxsz_loc = maxloc_result.loc;
+    *maxsz = (size_t) maxloc_result.sz;
+
+  fn_exit:
+    if (maxloc_type != MPI_DATATYPE_NULL)
+        MPIR_Type_free_impl(&maxloc_type);
+    if (maxloc_op != MPI_OP_NULL)
+        MPIR_Op_free_impl(&maxloc_op);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+static int shm_alloc_symm_all(MPIR_Comm * comm_ptr, size_t len, size_t off, void **ptr)
+{
+    int mpi_errno = MPI_SUCCESS, mpl_err = 0;
+    int rank = comm_ptr->rank;
+    int all_map_result = MPIDU_SYMSHM_MAP_FAIL;
+    int iter = MPIR_CVAR_SHM_SYMHEAP_RETRY;
+    int maxsz_loc = 0;
+    size_t maxsz = 0;
+    char *map_pointer = NULL;
+    MPIR_Comm *shm_comm_ptr = comm_ptr->node_comm;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    MPIDU_shm_seg_t *memory = NULL;
+    memory_list_t *memory_node = NULL;
+    MPIR_CHKPMEM_DECL(2);
+
+    mpi_errno = allreduce_maxloc(len, rank, comm_ptr, &maxsz, &maxsz_loc);
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+
+    if (maxsz == 0)
+        goto fn_exit;
+
+    MPIR_CHKPMEM_MALLOC(memory, MPIDU_shm_seg_t *, sizeof(*memory), mpi_errno, "memory_handle",
+                        MPL_MEM_OTHER);
+    MPIR_CHKPMEM_MALLOC(memory_node, memory_list_t *, sizeof(*memory_node), mpi_errno,
+                        "memory_node", MPL_MEM_OTHER);
+
+    mpl_err = MPL_shm_hnd_init(&(memory->hnd));
+    MPIR_ERR_CHKANDJUMP(mpl_err, mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+
+    memory->segment_len = len;
+
+    while (all_map_result == MPIDU_SYMSHM_MAP_FAIL && iter-- > 0) {
+        int map_result = MPIDU_SYMSHM_SUCCESS;
+
+        /* rank 0 in comm generates random address */
+        if (comm_ptr->rank == maxsz_loc)
+            map_pointer = generate_random_addr(len);
+
+        /* broadcast fixed address to the other processes in win */
+        mpi_errno =
+            MPIR_Bcast(&map_pointer, sizeof(char *), MPI_CHAR, maxsz_loc, comm_ptr, &errflag);
+        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+        /* adjust my base address */
+        memory->base_addr = map_pointer - off;
+
+        if (len > 0) {
+            if (shm_comm_ptr) {
+                if (shm_comm_ptr->rank == 0) {
+                    char *serialized_hnd = NULL;
```

Comment:
Makes sense

Reply:
I actually prefer the old code compared to the new version. This is because when I read the global retrial logic, I do not need to deal with the complex code for shm allocation:-)

Reply:
I have simplified the new code. Hopefully it reads better now. Now running tests on jenkins. Will perform more tests after these pass to check additional cases as you previously suggested.